### ISSUE TITLE
Add missing long error code explanations

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#61137]
This is a re-opening of #32777.

## Explanations

Those error codes and their explanations are declared in files named `error_codes.rs`. In those files, you can see generally two macros: `register_diagnostics` and `register_long_diagnostics`. If an error code is in `register_diagnostics`, it means it lacks long error explanation. The goal of this issue is to add those missing explanations!

This is an easy first issue but don't hesitate to ask for help if you're stuck on something!

Here is the list of **all** error codes that don't have a long error explanation yet:

 * [x] E0183 (#89885)
 * [x] E0203 (#66880)
 * [x] E0208 (#98011)
 * [x] E0212
 * [x] E0224
 * [x] E0226
 * [x] E0227 (#93809)
 * [x] E0228
 * [x] E0279 (not emitted anymore)
 * [x] E0280 (#106521)
 * [x] E0284
 * [x] E0307 (#64110)
 * [x] E0311 (https://github.com/rust-lang/rust/pull/100747)
 * [x] E0312 (#64347)
 * [x] E0313 (#106580)
 * [x] E0316
 * [x] E0320 (https://github.com/rust-lang/rust/pull/105792)
 * [x] E0377 (https://github.com/rust-lang/rust/pull/105935)
 * [x] E0457 (#105860)
 * [x] E0460 (#105860)
 * [x] E0461 (https://github.com/rust-lang/rust/pull/106028)
 * [ ] E0462
 * [x] E0464 (#89587)
 * [x] E0465 (https://github.com/rust-lang/rust/pull/106282)
 * [x] E0472 (#105791)
 * [x] E0476 (#108444)
 * [x] E0477
 * [x] E0482 (#89710)
 * [x] E0490 (https://github.com/rust-lang/rust/pull/106714)
 * [x] E0493 (#64377)
 * [x] E0495 (https://github.com/rust-lang/rust/pull/64404)
 * [x] E0498 (#87521)
 * [x] E0514 (https://github.com/rust-lang/rust/pull/106236)
 * [x] E0519 (https://github.com/rust-lang/rust/pull/106236)
 * [x] E0521
 * [ ] E0523
 * [x] E0524 (#64428)
 * [x] E0531 (#64455)
 * [x] E0533 (https://github.com/rust-lang/rust/pull/64735)
 * [x] E0539 (error than can only appear inside libstd code)
 * [x] E0542 (error than can only appear inside libstd code)
 * [x] E0543 (error than can only appear inside libstd code)
 * [x] E0544 (error than can only appear inside libstd code)
 * [x] E0545 (error than can only appear inside libstd code)
 * [x] E0546 (error than can only appear inside libstd code)
 * [x] E0547 (error than can only appear inside libstd code)
 * [x] E0549 (error than can only appear inside libstd code)
 * [x] E0550 (https://github.com/rust-lang/rust/pull/64838)
 * [x] E0551 (https://github.com/rust-lang/rust/pull/64918)
 * [x] E0556 (https://github.com/rust-lang/rust/pull/65055)
 * [x] E0561 (#65098)
 * [x] E0564 (#65135) not emitted anymore
 * [x] E0566 (#65164)
 * [x] E0567 (https://github.com/rust-lang/rust/pull/65179)
 * [x] E0568 (#65205)
 * [x] E0573 (#65234)
 * [x] E0574 (#65308)
 * [x] E0575 (#65334)
 * [x] E0576 (#65390)
 * [x] E0577 (#65434)
 * [x] E0578 (#65471)
 * [x] E0584 (https://github.com/rust-lang/rust/pull/65493)
 * [x] E0587 (#65563)
 * [x] E0588 (#65591)
 * [x] E0592 (#61822)
 * [x] E0594 (#66155)
 * [x] E0623 (#66186)
 * [x] E0625 (#87715)
 * [x] E0627
 * [x] E0628 (#69867)
 * [x] E0631 (#66979)
 * [x] E0632
 * [x] E0634 (#69998)
 * [x] E0637 (#68908)
 * [x] E0640 (#98011)
 * [x] E0641
 * [x] E0657
 * [x] E0666
 * [x] E0667 (https://github.com/rust-lang/rust/pull/95364)
 * [x] E0687
 * [x] E0688
 * [x] E0693 (#69993)
 * [x] E0696
 * [x] E0697 (#65215)
 * [x] E0703
 * [x] E0706
 * [x] E0708
 * [x] E0710
 * [x] E0711 (#98011)
 * [x] E0717 (#98011)
 * [x] E0719 (#69620)
 * [x] E0722
 * [x] E0724
 * [x] E0726
 * [x] E0727
 * [x] E0728
 * [x] E0734 (https://github.com/rust-lang/rust/pull/64763)
 * [x] E0739 (#69876)
 * [x] E0740 (#65405)
 * [x] E0772

**Comment by imbrem** [comment]
Can I take one of these on?

**Comment by GuillaumeGomez** [comment]
@imbrem: Sure, go ahead! When you open a PR, please link to this issue so others can see what's being worked on.

**Comment by imbrem** [comment]
Done! Actually, should I have requested/should I request you as a reviewer?

**Comment by Knium** [comment]
Can I take one of these on too?

**Comment by GuillaumeGomez** [comment]
Sure, go ahead (and thanks!). Just like I told previously: don't forget to link to this issue once your PR is open so we can keep track. :)

**Comment by estebank** [comment]
E0307 handled in #64110.

**Comment by JohnTitor** [comment]
@GuillaumeGomez Could you make E0697 checked?

**Comment by JohnTitor** [comment]
@GuillaumeGomez could you make E0666 and E0728 checked? It's done in #65855 and #65678.

**Comment by estebank** [comment]
@JohnTitor done

**Comment by smokytheangel0** [comment]
I was looking into E0183 and I cannot find where any explanation for it exists at all.
Do I have to try and make each error happen myself, in order to document it, or is there something I am missing entirely?

**Comment by JohnTitor** [comment]
@GuillaumeGomez E0623 should point to #66186, not #66018

**Comment by aDotInTheVoid** [comment]
How would I do this now that each error is in it's own `.md` file after #66314

**Comment by GuillaumeGomez** [comment]
The same but putting it into its own `.md` file?

**Comment by aDotInTheVoid** [comment]
E0203 handled by #66880

**Comment by reese** [comment]
I'd love to take on E0631! Will link back here in the PR once it's done.

**Comment by jwhite927** [comment]
I would like to address E0637.

**Comment by GuillaumeGomez** [comment]
@jwhite927 Sure, go ahead!

**Comment by thekuom** [comment]
Working on E0719 right now!

**Comment by ayushmishra2005** [comment]
@GuillaumeGomez I am picking  E0628. I will link PR, once it is done.

**Comment by ayushmishra2005** [comment]
@GuillaumeGomez I am picking E0693. I will link PR, once it is done.

**Comment by ayushmishra2005** [comment]
@GuillaumeGomez next I am picking E0634. I will link PR, once it is done.

**Comment by GuillaumeGomez** [comment]
@ayushmishra2005 Sorry, completely fell behind. Updating the issue. Thanks a lot for working on this issue! :)

**Comment by PankajChaudhary5** [comment]
I am picking E0657

**Comment by PankajChaudhary5** [comment]
I am picking E0696

**Comment by unexge** [comment]
[PR](https://github.com/rust-lang/rust/pull/71808) for E0539

**Comment by mbartlett21** [comment]
What happened to `--explain errors` from [RFC 1644]?

[RFC 1644]: https://rust-lang.github.io/rfcs/1644-default-and-expanded-rustc-errors.html

**Comment by GuillaumeGomez** [comment]
Still working for me.

**Comment by mbartlett21** [comment]
Passing `--explain errors` to rustc when compiling code says that it is not a recognised error code.

**Comment by GuillaumeGomez** [comment]
Ah no, my bad, I confused it with something else. I didn't hear about any development on that...

**Comment by mbartlett21** [comment]
It looks like issue #34827 was meant to be tracking that but it was closed in favour of this one.

**Comment by midgleyc** [comment]
Can I pick up E0722?

**Comment by GuillaumeGomez** [comment]
Sure, go ahead!

**Comment by GuillaumeGomez** [comment]
I updated the list (removed non-emitted errors as well).

**Comment by GuillaumeGomez** [comment]
No problem, go ahead!

**Comment by cameron1024** [comment]
Hi, can I work on E0183? :-)

**Comment by GuillaumeGomez** [comment]
Go ahead!

**Comment by cameron1024** [comment]
BTW, I had a look at some of the other PRs, and you've given some formatting feedback. Is there a "style guide" somewhere I should be following? Or is this something `./x.py check` would be able to catch?

**Comment by cameron1024** [comment]
Got it, thanks!

**Comment by kimjoaoun** [comment]
Hi everyone, can I work on `E0726`?

**Comment by GuillaumeGomez** [comment]
Go ahead!

**Comment by GuillaumeGomez** [comment]
Hum actually there is already https://github.com/rust-lang/rust/pull/87655 open... The author didn't fix the issues though but we can consider it as taken. Please pick another one.

**Comment by kimjoaoun** [comment]
That's sad, so can I work on `E0523`?

**Comment by GuillaumeGomez** [comment]
Seems like this one is free so go ahead!

**Comment by TmLev** [comment]
Can I claim `E0227`?

**Comment by GuillaumeGomez** [comment]
No problem (and you already sent the PR so perfect :wink: ).

**Comment by onlineSoftwareDevOK** [comment]
I am working on creating documentation for E0208. I need help. My first question is: How would you describe the error in a few words? Variance error? It seems it is related to types, traits, regions, object types, associated types and  associated consts

**Comment by mbartlett21** [comment]
@onlineSoftwareDevOK It seems to be an internal way to get tests for variance:

For example: this is one of its tests:

https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/variance/variance-regions-direct.stderr#L1-L9

**Comment by onlineSoftwareDevOK** [comment]
@mbartlett21 Thanks for your answer. That logic compiles. Where is the error? What am I missing?

**Comment by mbartlett21** [comment]
@onlineSoftwareDevOK
In other words, it isn't an error that would be presented to end users, since they have to have `#![feature(rustc_attrs)]` for it to appear. Presumably the documentation could just have a table mapping the symbol shown to the variance:

Symbol | Variance
-------|----------
`-`    | contravariant
`+`    | covariant
`o`    | invariant
`*`    | bivariant

**Comment by onlineSoftwareDevOK** [comment]
@mbartlett21 Thanks again. Why am I not able to replicate error E0208? Here's the output I get

```
warning: struct is never constructed: `Test2`
 --> test.rs:3:8
  |
3 | struct Test2<'a, 'b, 'c> {
  |        ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted

```

Code is:

```
#![feature(rustc_attrs)]

struct Test2<'a, 'b, 'c> {
     x: &'a isize,
     y: &'b [isize],
     c: &'c str
}

fn main() {}

```

**Comment by mbartlett21** [comment]
Copy the test, specifically line 8:

https://github.com/rust-lang/rust/blob/673d0db5e393e9c64897005b470bfeb6d5aec61b/src/test/ui/variance/variance-regions-direct.rs#L6-L13

**Comment by onlineSoftwareDevOK** [comment]
@mbartlett21 Thank you, it worked. Can you explain how the correct version of this code would look like, please? I guess this is a contravariant error (since I am getting three - ) but I am not sure

**Comment by mbartlett21** [comment]
@onlineSoftwareDevOK 
Perhaps you didn't understand properly. It is not an error that needs to be fixed as such. It only exists so that variance can be tested with the rest of the test suite.

See for example:
https://github.com/rust-lang/rust/blob/b97dc20784698730bd7bc1a20eabdcd32066d174/src/test/ui/feature-gates/feature-gate-rustc-attrs-1.rs#L3
https://github.com/rust-lang/rust/blob/b97dc20784698730bd7bc1a20eabdcd32066d174/src/test/ui/feature-gates/feature-gate-rustc-attrs-1.stderr#L1-L7
https://github.com/rust-lang/rust/blob/f6e5570460b2bb925021dc667ead2a9834ea88cb/compiler/rustc_typeck/src/variance/test.rs#L5-L14

**Comment by onlineSoftwareDevOK** [comment]
@mbartlett21 Thanks for clarifying. I will prepare the documentation and create a PR in the next days

**Comment by onlineSoftwareDevOK** [comment]
I am preparing documentation for E0279. If I compile this:

```
fn main() {
    test::<FooS>(&mut 42);
}

trait Foo<'a> {}

struct FooS<'a> {
    data: &'a mut u32,
}

impl<'a, 'b: 'a> Foo<'b> for FooS<'a> {}

fn test<'a, F>(data: &'a mut u32) where F: for<'b> Foo<'b> {}
```

I get this:

```
error: implementation of `Foo` is not general enough
 --> test.rs:2:5
  |
2 |     test::<FooS>(&mut 42);
  |     ^^^^^^^^^^^^ implementation of `Foo` is not general enough
  |
  = note: `FooS<'_>` must implement `Foo<'0>`, for any lifetime `'0`...
  = note: ...but `FooS<'_>` actually implements `Foo<'1>`, for some specific lifetime `'1`

error: aborting due to previous error

```

The following links seem to explain this: https://github.com/rust-lang/rust/issues/35180#issue-168838480, https://github.com/rust-lang/rust/issues/35180#issuecomment-526994483, https://github.com/rust-lang/rust/pull/64085

My question is: Does E0279 still exist? I tried replicating it but it seems it does not happen anymore. If it still exists, can you please post an example here to replicate the error, please?

**Comment by GuillaumeGomez** [comment]
`E0279` is emitted from `compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs` apparently.

**Comment by TmLev** [comment]
E0667 can be marked as done: https://github.com/rust-lang/rust/commit/eae29031504d62769b94dffca4aba3d78a6fafae

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by mdholloway** [comment]
> My question is: Does E0279 still exist? I tried replicating it but it seems it does not happen anymore. If it still exists, can you please post an example here to replicate the error, please?

> `E0279` is emitted from `compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs` apparently.

Not anymore as of #99501.  Can't tell whether it's permanently or only temporarily gone, though.

**Comment by pnkfelix** [comment]
Visited during T-compiler [backlog bonanza](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.20meeting.5D.202022-08-12.20compiler-team.23484.20backlog.20b'za/near/293119423).

I believe this is more accurately a metabug rather than a tracking issue. (If we *were* going to tag it with some S-tracking-* label, then I guess it would be S-tracking-impl-incomplete, but really, its not like this is a feature we're implementing. Its just a most excellent list of work items.)

@rustbot label: -C-tracking-issue +metabug

**Comment by MatthewPeterKelly** [comment]
It looks like `E0772` has a detailed error description linked in `error_codes.rs`, but is marked as "open" in the description for this issue. The long description in `compiler/rustc_error_codes/src/error_codes/E0772.md` also notes that this error is no longer emitted by the compiler.

**Comment by GuillaumeGomez** [comment]
I checked it.

**Comment by akabinds** [comment]
I was working on adding an explanation for E0313. But, I hit a roadblock in doing so and I was recommended by other contributors to propose a deprecation for the error code (#103742) since it seemingly isn't emitted anymore by the current revision of the borrow checker and it has no UI tests. Please view the reference issue for more background.

**Comment by Ezrashaw** [comment]
@GuillaumeGomez *sigh*, turns out that `E0476` *isn't* unreachable, I'll have to document it instead.

**Comment by GuillaumeGomez** [comment]
It's fine. I'll update the list once the new PR is open.
